In [2]:
import pandas as pd
from sklearn.cluster import KMeans

In [18]:
df_test=pd.read_csv("./datasets/data_set_ALL_AML_independent.csv")
df_train=pd.read_csv("./datasets/data_set_ALL_AML_train.csv")
y=pd.read_csv("./datasets/actual.csv")
print(df_test.shape)
print(df_train.shape)
print(y.shape)

(7129, 70)
(7129, 78)
(72, 2)


In [22]:
df_test.head()

Gene Description Gene Accession Number   39 call   40  \
0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -342    A  -87   
1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -200    A -248   
2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   41    A  262   
3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at  328    A  295   
4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -224    A -226   

  call.1   42 call.2   47 call.3  ...   65 call.29   66 call.30   63 call.31  \
0      A   22      A -243      A  ...  -62       A  -58       A -161       A   
1      A -153      A -218      A  ... -198       A -217       A -215       A   
2      A   17      A -163      A  ...   -5       A   63       A  -46       A   
3      A  276      A  182      A  ...  141       A   95       A  146       A   
4      A -211      A -289      A  ... -256       A -191       A -172       A   

    64 call.32   62 call.33  
0  -48       A -176       A  
1 -531       A -284       A  
2 -124       A  -81       A  
3  431       A    9       A  
4 -496       A -294       A  

[5 rows x 70 columns]

In [23]:
df_train.head()

Gene Description Gene Accession Number    1 call    2  \
0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -214    A -139   
1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -153    A  -73   
2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at  -58    A   -1   
3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at   88    A  283   
4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -295    A -264   

  call.1    3 call.2    4 call.3  ...   29 call.33   30 call.34   31 call.35  \
0      A  -76      A -135      A  ...   15       A -318       A  -32       A   
1      A  -49      A -114      A  ... -114       A -192       A  -49       A   
2      A -307      A  265      A  ...    2       A  -95       A   49       A   
3      A  309      A   12      A  ...  193       A  312       A  230       P   
4      A -376      A -419      A  ...  -51       A -139       A -367       A   

    32 call.36   33 call.37  
0 -124       A -135       A  
1  -79       A -186       A  
2  -37       A  -70       A  
3  330       A  337       A  
4 -188       A -407       A  

[5 rows x 78 columns]

In [24]:
y.head()

patient  cancer
0        1       0
1        2       0
2        3       0
3        4       0
4        5       0

In [25]:
y["cancer"].value_counts()

0    47
1    25
Name: cancer, dtype: int64

In [26]:
y = y.replace({'ALL':0,'AML':1})
y

patient  cancer
0         1       0
1         2       0
2         3       0
3         4       0
4         5       0
..      ...     ...
67       68       0
68       69       0
69       70       0
70       71       0
71       72       0

[72 rows x 2 columns]

#### PreProcessing

In [27]:
# Transform all the call values to numbers (not used in this version)
# df_train.replace(['A','P','M'],['1','2','3'], inplace=True)
# df_test.replace(['A','P','M'],['1','2','3'], inplace=True)

# Remove "call" columns from training and testing data
train_to_keep = [col for col in df_train.columns if "call" not in col]
test_to_keep = [col for col in df_test.columns if "call" not in col]

X_train_tr = df_train[train_to_keep]
X_test_tr = df_test[test_to_keep]

In [28]:
X_test_tr.head()

Gene Description Gene Accession Number   39   40   42  \
0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -342  -87   22   
1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -200 -248 -153   
2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   41  262   17   
3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at  328  295  276   
4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -224 -226 -211   

    47   48   49   41   43  ...   54   57   58   60   61   65   66   63   64  \
0 -243 -130 -256  -62   86  ...  -90 -137 -157 -172  -47  -62  -58 -161  -48   
1 -218 -177 -249  -23  -36  ...  -87  -51 -370 -122 -442 -198 -217 -215 -531   
2 -163  -28 -410   -7 -141  ...  102  -82  -77   38  -21   -5   63  -46 -124   
3  182  266   24  142  252  ...  319  178  340   31  396  141   95  146  431   
4 -289 -170 -535 -233 -201  ... -283 -135 -438 -201 -351 -256 -191 -172 -496   

    62  
0 -176  
1 -284  
2  -81  
3    9  
4 -294  

[5 rows x 36 columns]

In [29]:
X_train_tr

Gene Description Gene Accession Number  \
0                   AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at   
1                   AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at   
2                   AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   
3                   AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at   
4                   AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at   
...                                                 ...                   ...   
7124  PTGER3 Prostaglandin E receptor 3 (subtype EP3...             X83863_at   
7125  HMG2 High-mobility group (nonhistone chromosom...             Z17240_at   
7126      RB1 Retinoblastoma 1 (including osteosarcoma)           L49218_f_at   
7127  GB DEF = Glycophorin Sta (type A) exons 3 and ...           M71243_f_at   
7128                          GB DEF = mRNA (clone 1A7)           Z78285_f_at   

        1    2     3    4    5    6     7     8  ...    35    36    37    38  \
0    -214 -139   -76 -135 -106 -138   -72  -413  ...     7  -213   -25   -72   
1    -153  -73   -49 -114 -125  -85  -144  -260  ...  -100  -252   -20  -139   
2     -58   -1  -307  265  -76  215   238     7  ...   -57   136   124    -1   
3      88  283   309   12  168   71    55    -2  ...   132   318   325   392   
4    -295 -264  -376 -419 -230 -272  -399  -541  ...  -377  -209  -396  -324   
...   ...  ...   ...  ...  ...  ...   ...   ...  ...   ...   ...   ...   ...   
7124  793  782  1138  627  250  645  1140  1799  ...   752  1293  1733  1567   
7125  329  295   777  170  314  341   482   446  ...   295   342   304   627   
7126   36   11    41  -50   14   26    10    59  ...    28    26    12    21   
7127  191   76   228  126   56  193   369   781  ...  1558   246  3193  2520   
7128  -37  -14   -41  -91  -25  -53   -42    20  ...   -67    23   -33     0   

       28    29   30   31    32    33  
0      -4    15 -318  -32  -124  -135  
1    -116  -114 -192  -49   -79  -186  
2    -125     2  -95   49   -37   -70  
3     241   193  312  230   330   337  
4    -191   -51 -139 -367  -188  -407  
...   ...   ...  ...  ...   ...   ...  
7124  987   279  737  588  1170  2315  
7125  279    51  227  361   284   250  
7126   22     6   -9  -26    39   -12  
7127  662  2484  371  133   298   790  
7128  -46    -2  -31  -32    -3   -10  

[7129 rows x 40 columns]

In [31]:
print(X_train_tr.shape)
print(X_test_tr.shape)

(7129, 40)
(7129, 36)


In [33]:
train_columns_titles = ['Gene Description', 'Gene Accession Number', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
       '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', 
       '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38']

X_train_tr = X_train_tr.reindex(columns=train_columns_titles)
test_columns_titles = ['Gene Description', 'Gene Accession Number','39', '40', '41', '42', '43', '44', '45', '46',
       '47', '48', '49', '50', '51', '52', '53',  '54', '55', '56', '57', '58', '59',
       '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72']

X_test_tr = X_test_tr.reindex(columns=test_columns_titles)

In [34]:
X_train_tr.head()

Gene Description Gene Accession Number    1    2    3  \
0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -214 -139  -76   
1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -153  -73  -49   
2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at  -58   -1 -307   
3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at   88  283  309   
4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -295 -264 -376   

     4    5    6    7    8  ...   29   30   31   32   33   34   35   36   37  \
0 -135 -106 -138  -72 -413  ...   15 -318  -32 -124 -135  -20    7 -213  -25   
1 -114 -125  -85 -144 -260  ... -114 -192  -49  -79 -186 -207 -100 -252  -20   
2  265  -76  215  238    7  ...    2  -95   49  -37  -70  -50  -57  136  124   
3   12  168   71   55   -2  ...  193  312  230  330  337  101  132  318  325   
4 -419 -230 -272 -399 -541  ...  -51 -139 -367 -188 -407 -369 -377 -209 -396   

    38  
0  -72  
1 -139  
2   -1  
3  392  
4 -324  

[5 rows x 40 columns]

In [35]:
X_test_tr

Gene Description Gene Accession Number  \
0                   AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at   
1                   AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at   
2                   AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   
3                   AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at   
4                   AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at   
...                                                 ...                   ...   
7124  PTGER3 Prostaglandin E receptor 3 (subtype EP3...             X83863_at   
7125  HMG2 High-mobility group (nonhistone chromosom...             Z17240_at   
7126      RB1 Retinoblastoma 1 (including osteosarcoma)           L49218_f_at   
7127  GB DEF = Glycophorin Sta (type A) exons 3 and ...           M71243_f_at   
7128                          GB DEF = mRNA (clone 1A7)           Z78285_f_at   

        39   40   41    42    43   44   45   46  ...   63   64   65    66  \
0     -342  -87  -62    22    86 -146 -187  -56  ... -161  -48  -62   -58   
1     -200 -248  -23  -153   -36  -74 -187  -43  ... -215 -531 -198  -217   
2       41  262   -7    17  -141  170  312   43  ...  -46 -124   -5    63   
3      328  295  142   276   252  174  142  177  ...  146  431  141    95   
4     -224 -226 -233  -211  -201  -32  114 -116  ... -172 -496 -256  -191   
...    ...  ...  ...   ...   ...  ...  ...  ...  ...  ...  ...  ...   ...   
7124  1074   67  245   893  1235  354  304  625  ...  809  466  707   423   
7125   475  263  164   297     9  -42   -1  173  ...  445  349  354    41   
7126    48  -33   84     6     7 -100 -207   63  ...   -2    0  -22     0   
7127   168  -33  100  1971  1545   45  112   63  ...  210  284  260  1777   
7128   -70  -21  -18   -42   -81 -108 -190  -62  ...   16  -73    5   -49   

       67   68   69   70   71   72  
0     -76 -154  -79  -55  -59 -131  
1     -98 -136 -118  -44 -114 -126  
2    -153   49  -30   12   23  -50  
3     237  180   68  129  146  211  
4    -215 -257 -110 -108 -171 -206  
...   ...  ...  ...  ...  ...  ...  
7124  441  524  742  320  348  874  
7125   99  249  234  174  208  393  
7126   -8   40   72   -4    0   34  
7127   80  -68  109  176   74  237  
7128  -12   -1  -30   40  -12   -2  

[7129 rows x 36 columns]

In [36]:
X_train = X_train_tr.T
X_test = X_test_tr.T


In [37]:
X_train.head()

0     \
Gene Description       AFFX-BioB-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-5_at   
1                                                     -214   
2                                                     -139   
3                                                      -76   

                                                      1     \
Gene Description       AFFX-BioB-M_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-M_at   
1                                                     -153   
2                                                      -73   
3                                                      -49   

                                                      2     \
Gene Description       AFFX-BioB-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-3_at   
1                                                      -58   
2                                                       -1   
3                                                     -307   

                                                      3     \
Gene Description       AFFX-BioC-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioC-5_at   
1                                                       88   
2                                                      283   
3                                                      309   

                                                      4     \
Gene Description       AFFX-BioC-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioC-3_at   
1                                                     -295   
2                                                     -264   
3                                                     -376   

                                                       5     \
Gene Description       AFFX-BioDn-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioDn-5_at   
1                                                      -558   
2                                                      -400   
3                                                      -650   

                                                       6     \
Gene Description       AFFX-BioDn-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioDn-3_at   
1                                                       199   
2                                                      -330   
3                                                        33   

                                                      7     \
Gene Description       AFFX-CreX-5_at (endogenous control)   
Gene Accession Number                       AFFX-CreX-5_at   
1                                                     -176   
2                                                     -168   
3                                                     -367   

                                                      8     \
Gene Description       AFFX-CreX-3_at (endogenous control)   
Gene Accession Number                       AFFX-CreX-3_at   
1                                                      252   
2                                                      101   
3                                                      206   

                                                      9     ...  \
Gene Description       AFFX-BioB-5_st (endogenous control)  ...   
Gene Accession Number                       AFFX-BioB-5_st  ...   
1                                                      206  ...   
2                                                       74  ...   
3                                                     -215  ...   

                                                            7119  \
Gene Description       Transcription factor Stat5b (stat5b) mRNA   
Gene Accession Number                                  U48730_at   
1                                                            185   
2         

In [38]:
X_test.head()

0     \
Gene Description       AFFX-BioB-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-5_at   
39                                                    -342   
40                                                     -87   
41                                                     -62   

                                                      1     \
Gene Description       AFFX-BioB-M_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-M_at   
39                                                    -200   
40                                                    -248   
41                                                     -23   

                                                      2     \
Gene Description       AFFX-BioB-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-3_at   
39                                                      41   
40                                                     262   
41                                                      -7   

                                                      3     \
Gene Description       AFFX-BioC-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioC-5_at   
39                                                     328   
40                                                     295   
41                                                     142   

                                                      4     \
Gene Description       AFFX-BioC-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioC-3_at   
39                                                    -224   
40                                                    -226   
41                                                    -233   

                                                       5     \
Gene Description       AFFX-BioDn-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioDn-5_at   
39                                                     -427   
40                                                     -493   
41                                                     -284   

                                                       6     \
Gene Description       AFFX-BioDn-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioDn-3_at   
39                                                     -656   
40                                                      367   
41                                                     -167   

                                                      7     \
Gene Description       AFFX-CreX-5_at (endogenous control)   
Gene Accession Number                       AFFX-CreX-5_at   
39                                                    -292   
40                                                    -452   
41                                                     -97   

                                                      8     \
Gene Description       AFFX-CreX-3_at (endogenous control)   
Gene Accession Number                       AFFX-CreX-3_at   
39                                                     137   
40                                                     194   
41                                                     -12   

                                                      9     ...  \
Gene Description       AFFX-BioB-5_st (endogenous control)  ...   
Gene Accession Number                       AFFX-BioB-5_st  ...   
39                                                    -144  ...   
40                                                     162  ...   
41                                                     -70  ...   

                                                            7119  \
Gene Description       Transcription factor Stat5b (stat5b) mRNA   
Gene Accession Number                                  U48730_at   
39                                                           277   
40        

In [39]:
# Clean up the column names for training and testing data
X_train.columns = X_train.iloc[1]
X_train = X_train.drop(["Gene Description", "Gene Accession Number"]).apply(pd.to_numeric)

# Clean up the column names for Testing data
X_test.columns = X_test.iloc[1]
X_test = X_test.drop(["Gene Description", "Gene Accession Number"]).apply(pd.to_numeric)

In [40]:
X_train.head()

Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \
1                                -214            -153             -58   
2                                -139             -73              -1   
3                                 -76             -49            -307   
4                                -135            -114             265   
5                                -106            -125             -76   

Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \
1                                  88            -295             -558   
2                                 283            -264             -400   
3                                 309            -376             -650   
4                                  12            -419             -585   
5                                 168            -230             -284   

Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \
1                                  199            -176             252   
2                                 -330            -168             101   
3                                   33            -367             206   
4                                  158            -253              49   
5                                    4            -122              70   

Gene Accession Number  AFFX-BioB-5_st  ...  U48730_at  U58516_at  U73738_at  \
1                                 206  ...        185        511       -125   
2                                  74  ...        169        837        -36   
3                                -215  ...        315       1199         33   
4                                  31  ...        240        835        218   
5                                 252  ...        156        649         57   

Gene Accession Number  X06956_at  X16699_at  X83863_at  Z17240_at  \
1                            389        -37        793        329   
2                            442        -17        782        295   
3                            168         52       1138        777   
4                            174       -110        627        170   
5                            504        -26        250        314   

Gene Accession Number  L49218_f_at  M71243_f_at  Z78285_f_at  
1                               36          191          -37  
2                               11           76          -14  
3                               41          228          -41  
4                              -50          126          -91  
5                               14           56          -25  

[5 rows x 7129 columns]

In [41]:
X_test.head()

Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \
39                               -342            -200              41   
40                                -87            -248             262   
41                                -62             -23              -7   
42                                 22            -153              17   
43                                 86             -36            -141   

Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \
39                                328            -224             -427   
40                                295            -226             -493   
41                                142            -233             -284   
42                                276            -211             -250   
43                                252            -201             -384   

Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \
39                                -656            -292             137   
40                                 367            -452             194   
41                                -167             -97             -12   
42                                  55            -141               0   
43                                -420            -197             -60   

Gene Accession Number  AFFX-BioB-5_st  ...  U48730_at  U58516_at  U73738_at  \
39                               -144  ...        277       1023         67   
40                                162  ...         83        529       -295   
41                                -70  ...        129        383         46   
42                                500  ...        413        399         16   
43                               -468  ...        341         91        -84   

Gene Accession Number  X06956_at  X16699_at  X83863_at  Z17240_at  \
39                           214       -135       1074        475   
40                           352        -67         67        263   
41                           104         15        245        164   
42                           558         24        893        297   
43                           615        -52       1235          9   

Gene Accession Number  L49218_f_at  M71243_f_at  Z78285_f_at  
39                              48          168          -70  
40                             -33          -33          -21  
41                              84          100          -18  
42                               6         1971          -42  
43                               7         1545          -81  

[5 rows x 7129 columns]

That looks much better. We have the 38 patients as rows in the training set, and the other 34 as rows in the testing set. Each of those datasets has 7129 gene expression features.